remote/remote_state_gcs.go

```diff
@@ -0,0 +1,405 @@
+	terragruntOptions.Logger.Printf("Creating GCS bucket %s in project %s", config.remoteStateConfigGCS.Bucket, config.Project)
+
+	// The project ID to which the bucket belongs. This is only used when creating a new bucket during initialization.
+	// Since buckets have globally unique names, the project ID is not required to access the bucket during normal
+	// operation.
+	projectID := config.Project
+
+	ctx := context.Background()
+	bucket := gcsClient.Bucket(config.remoteStateConfigGCS.Bucket)
+
+	bucketAttrs := &storage.BucketAttrs{}
+
+	if config.Location != "" {
+		terragruntOptions.Logger.Printf("Creating GCS bucket in location %s.", config.Location)
+	}
+
+	if config.SkipBucketVersioning {
+		terragruntOptions.Logger.Printf("Versioning is disabled for the remote state GCS bucket %s using 'skip_bucket_versioning' config.", config.remoteStateConfigGCS.Bucket)
+	} else {
+		terragruntOptions.Logger.Printf("Enabling versioning on GCS bucket %s", config.remoteStateConfigGCS.Bucket)
+		bucketAttrs.VersioningEnabled = true
+	}
+
+	err := bucket.Create(ctx, projectID, bucketAttrs)
+	return errors.WithStackTrace(err)
+}
+
+// GCP is eventually consistent, so after creating a GCS bucket, this method can be used to wait until the information
+// about that GCS bucket has propagated everywhere.
+func WaitUntilGCSBucketExists(gcsClient *storage.Client, config *RemoteStateConfigGCS, terragruntOptions *options.TerragruntOptions) error {
+	terragruntOptions.Logger.Printf("Waiting for bucket %s to be created", config.Bucket)
+	for retries := 0; retries < MAX_RETRIES_WAITING_FOR_GCS_BUCKET; retries++ {
+		if DoesGCSBucketExist(gcsClient, config) {
+			terragruntOptions.Logger.Printf("GCS bucket %s created.", config.Bucket)
+			return nil
+		} else if retries < MAX_RETRIES_WAITING_FOR_GCS_BUCKET-1 {
+			terragruntOptions.Logger.Printf("GCS bucket %s has not been created yet. Sleeping for %s and will check again.", config.Bucket, SLEEP_BETWEEN_RETRIES_WAITING_FOR_GCS_BUCKET)
+			time.Sleep(SLEEP_BETWEEN_RETRIES_WAITING_FOR_GCS_BUCKET)
+		}
+	}
+
+	return errors.WithStackTrace(MaxRetriesWaitingForS3BucketExceeded(config.Bucket))
+}
+
+// DoesGCSBucketExist returns true if the GCS bucket specified in the given config exists and the current user has the
+// ability to access it.
+func DoesGCSBucketExist(gcsClient *storage.Client, config *RemoteStateConfigGCS) bool {
+	ctx := context.Background()
+
+	// Creates a Bucket instance.
+	bucket := gcsClient.Bucket(config.Bucket)
+
+	// TODO - the code below attempts to determine whether the storage bucket exists by making a making a number of API
+	// calls, then attemping to list the contents of the bucket. It was adapted from Google's own integration tests and
+	// should be improved once the appropriate API call is added. For more info see:
+	// https://github.com/GoogleCloudPlatform/google-cloud-go/blob/de879f7be552d57556875b8aaa383bce9396cc8c/storage/integration_test.go#L1231
```

Comment:
Thx for including these comments! Helps clarify why the code is so complicated for something that seems simple :)